Tracer and Tainted Registry Ability in DBA and TRACER, OBHOOK Updates

Fatgle opened the pull request:
## TRACER and OBHOOK 
1. TRACER allows callback function to pass an argument defined by user now.
2. OBHOOK allows callback function to pass an argument defined by user now.
3. Test cases of both of TRACER and OBHOOK have been revised.

## DBA
### The results of DBA tasks with tracer ability are expected to be written into files.
### Tracer ability
1. DBA module now has tracer ability as dba/dba_tracer.h and dba/dba_tracer.c.
### User Interface
- Tracer Ability
1. Questions of loading structure definition and global variable JSON files have been added to the DBA setting and DBA API.
- Taint Ability
1. The tainted packet question has been added to the DBA setting ( User should turn on NTM manually before this commit, but NTM will be turned on if it's needed now ).
2. The tainted registry has been added to the DBA.
### DBA Configuration
- dba_config_sample
1. The dba_config_sample has been updated according to this commit.
- Taint Ability
1. User can specify the files to write the captured packets in the tags under the "NETTRAMON".
2. "PACKET" and "TAG" will be ignored if the "ENABLE" tag of "TAINT" is "no".
- Tracer Ability
1. User can specify the files for structure definition and global variable JSON files under the "TRACER" tag.
2. "ENABLE" under "TRACER is the master switch of TRACER ability and config parser will ignore all the setting below if it is "no".
### DBA user interface setting sequence goes as:
**"TAINT"** ( -> "PACKET" -> "tag" ) -> **"SYSCALL"** -> **"TRACER"** [ ( -> "LOAD STRUCTURE" -> "LOAD GLOBALVAR" ) -> "INSTRUCTION" ( -> "IS_KERNEL" ) -> "BLOCK" ( -> "IS_KERNEL" ) ] -> "CONFIRM"
### DBA config file setting sequence goes as:
**"MODULES"** -> "AGENT" -> "DIFT" -> "NETTRAMON" ( -> "ALL" -> "TCP" -> "UDP" -> "ICMP" ) -> **"TAINT"** ( -> "PACKET" -> "tag" ) -> **"SYSCALL"** -> **"TRACER"** [ ( -> "LOAD STRUCTURE" -> "LOAD GLOBALVAR" ) -> "INSTRUCTION" ( -> "IS_KERNEL" ) -> "BLOCK" ( -> "IS_KERNEL" ) ] -> "CONFIRM"
## TSK 
1. Add qemu command for print registry value by address.
2. Revise MBA command spec. ex. get_file -> mba_get_file
3. Add taint registry value into dba
4. Improve efficiency about search taint file and registry of tainted address.
5. Parse registry log file & merge to primary file.
6. Support print HKCU path in disk, but not suppout found tainted on HKCU  yet.

Fatgle committed:
1. Add qemu command for print registry value by address.
2. Revise MBA command spec. ex. get_file -> mba_get_file

Fatgle committed:
Add taint registry value into dba

Fatgle committed:
Improve efficiency about search taint file and registry of tainted address.

JeffreyJao committed:
OBHOOK and DBA Module Updated

1. OBHOOK now allows user to transmit an argument of callback function
2. Move arguments of nettramon in DBA into taint structure
3. Revise test cases of OBHOOK due to the first edition
4. Add license

JeffreyJao committed:
TRACER Module Updated

1. Allow callback function to pass argument defined by user
2. Modify tracer test cases in order to build without warning

JeffreyJao committed:
DBA Module Updated

1. Added instruction tracer ability to DBA module as dba/dba_tracer.h and dba/dba_tracer.c
2. Made return value of function "target_disas_inst_count" become integer
3. Revised Makefile.target

JeffreyJao committed:
DBA Module Updated

1. Added the tainted packet question to the DBA setting ( User should turn on NTM manually before this commit, but NTM will be turned on if it's needed now ).
   1.1. DBA allows user to set the targeted file for writing captured packets together or separately according to different protocols.
2. Added loading structure definition and global variable json files questions to DBA setting and DBA API if tracer ability is needed.
   2.1. Load once for all tasks.
   2.2. Duplicated loading will be handled by MEMFRS since DBA has no global context.
3. Updated dba_config_sample and the parsing function
   3.1. The "ENABLE" tag of TRACER is the master switch, so both of "INSTR" and "BLOCK" will be turned off no matter what their "ENABLE" tag is set.
   3.2. "PACKET" tag will be read iff TAINT is needed, so if user want to monitor network traffic while no TAINT ability, filling the "NETTRAMON" tag in "MODULES".
4. Updated DBA setting question sequence.
   4.1. The sequence goes as : "TAINT" ( -> "PACKET" -> "tag" ) -> "SYSCALL" -> "TRACER" ( -> "LOAD STRUCTURE" -> "LOAD GLOBALVAR" )
                               -> "INSTRUCTION" ( -> "IS_KERNEL" ) -> "BLOCK" ( -> "IS_KERNEL" ) -> "CONFIRM"

JeffreyJao committed:
DBA Module Code Revise

1. dba-commands.c and dba-commands.h
   1.1. Remove dependence between dba_set_by_config and all the cb_dba_xxx functions used in setting via user interface.
   1.2. Remove do_dba_config_file_setting which is used in MBA/monitor.c
2. Make showing messages correctly while setting DBA via config file.
3. Revise comments.

JeffreyJao committed:
Systrace Module Added

Add Systrace module by MisterLiHao

Done:
- add systrace module

Note:
- race condition may occur when users add systrace callback functions

Some parts are still incomplete:
- built-in syscall parsing callback functions
- detail error messages

Fatgle committed:
parse registry log file & merge to primary file.

GlacierW commented on ext/dba/dba.h:
```diff
@@ -143,6 +175,88 @@ extern int dba_set_timer( DBA_TID tid, size_t seconds );
+///     \param  path    path to the global variable json file
+///
+/// Return 0 on success, otherwise -1 is returned 
+extern int dba_set_global( DBA_TID tid, const char* path );
```
Make the API naming more clear. 
e.g. `dba_set_globalvar_json( DBA_TID tid, const char* json_fpath )`
The "set_global" can be ambiguous.

GlacierW commented on ext/dba/dba.h:
```diff
@@ -143,6 +175,88 @@ extern int dba_set_timer( DBA_TID tid, size_t seconds );
+///     \param  path    path to the struction definition json file
+///
+/// Return 0 on success, otherwise -1 is returned 
+extern int dba_set_structure( DBA_TID tid, const char* path );
```
Similar to `dba_set_global`

GlacierW commented on ext/dba/dba_taint.c:
```diff
@@ -255,18 +307,249 @@ int enum_tainted_file( dba_context* ctx ) {
+    /// and recover the blocks to high-level file information
+    img = get_device_image( "ide0-hd0" );
+    if ( get_hive_file( "/Windows/System32/config/SAM", "./SAM") < 0 )    
+        printf("Download SAM failed\n");
```
avoid using printf.
use monitor_printf or error code return instead.

Fatgle committed:
1. Add recovery all HKLM hive log ex.SAM.LOG1 SAM.LOG2
2. Support print HKCU path
3. fix printf to monitor_printf in dba_taint.c
4. Chenge REGISTRY_MAX_LENGTH to 500

JeffreyJao committed:
DBA User Interface Code Revise

1. Fix the bug that users disable tracer but the information does not be updated.

Fatgle committed:
Make the DBA API naming more clear.

Fatgle committed:
Found tainted SAM, SOFTWARE, SYSTEM, SECURITY by their LOG1 and LOG2.

Fatgle committed:
1. Support HKCU -> NTUSER.DAT recovery by LOG.
2. Change found tainted registry policy.

Before
    recovery_hive_by_all_log;
Now
    if ( registry log tainted )
        recovery_hive_by_log;

Fatgle committed:
Fix bug of recovery log to primary hive.

GlacierW committed:
Merge branch 'master' into master
